### PR TITLE
add epub 3.x support

### PR DESCRIPTION
**Pull request by ag2s20150909**
add epub 3.x support

### CODE REVIEW

**Comment by jiakuan** [comment]
@ag2s20150909 Thanks for your pull request. There is a test failure at:
https://github.com/documentnode/epub4j/pull/3/checks?check_run_id=2192301875

Once that is fixed, I will have a deep look the code changes.

File: epub4j-core/src/main/java/io/documentnode/epub4j/domain/LazyResource.java [hunk 1]
```diff
@@ -1,7 +1,8 @@
 package io.documentnode.epub4j.domain;
 
+//import android.util.Log;
```
Comment (jiakuan) [inline epub4j-core/src/main/java/io/documentnode/epub4j/domain/LazyResource.java:3]: Can we remove unused imports? If possible, it'd be great to not depend on Android libraries in this library.

File: epub4j-core/src/main/java/io/documentnode/epub4j/domain/LazyResource.java [hunk 2]
```diff
@@ -13,7 +14,7 @@
 public class LazyResource extends Resource {
 
   private static final long serialVersionUID = 5089400472352002866L;
-  private static final Logger log = Logger.create(LazyResource.class);
+  private  final String TAG= getClass().getName();
```
Comment (jiakuan) [inline epub4j-core/src/main/java/io/documentnode/epub4j/domain/LazyResource.java:17]: This is unused, it seems we can remove it.

File: epub4j-core/src/main/java/io/documentnode/epub4j/domain/Resource.java [hunk 3]
```diff
@@ -23,9 +23,13 @@
...
   private String href;
-  protected String originalHref;
+
+
+
+  private String properties;
+  protected final String originalHref;
   private MediaType mediaType;
-  private String inputEncoding = Constants.CHARACTER_ENCODING;
+  private String inputEncoding;
```
Comment (jiakuan) [inline epub4j-core/src/main/java/io/documentnode/epub4j/domain/Resource.java:32]: Is there a default encoding for EPUB 3?

File: epub4j-core/src/test/java/io/documentnode/epub4j/epub/ResourcesLoaderTest.java [hunk 4]
```diff
@@ -14,175 +14,176 @@
...
-    resource = resources.getByHref(allHrefs.get(2));
-    Assert.assertEquals("chapter1", resource.getId());
-    Assert.assertEquals("OEBPS/chapter1.html", resource.getHref());
-    Assert.assertEquals(MediaTypes.XHTML, resource.getMediaType());
-    Assert.assertEquals(5100, resource.getData().length);
-    expectedData = IOUtil.toByteArray(
-        this.getClass().getResourceAsStream("/book1/chapter1.html"));
-    Assert.assertArrayEquals(expectedData, resource.getData());
-  }
+//  private static final String encoding = "UTF-8";
```
Comment (jiakuan) [inline epub4j-core/src/test/java/io/documentnode/epub4j/epub/ResourcesLoaderTest.java:28]: Shall we have tests still?

**Review by jiakuan — COMMENTED** [review]
Thanks for the pull request, which looks great.

Just need a bit cleanup for unused imports and commented lines. If possible, we should keep existing tests for EPUB 2, and will need more tests to make sure all changes regarding EPUB 3 is covered.

We had a `Sample.java` to generate EPUB 2 books so that we can validate against http://validator.idpf.org/ as integration tests. Can we please have a similar class like `SampleV3.java` for generating EPUB 3 files?

**Comment by ag2s20150909** [comment]
java.lang.AssertionError: array lengths differed, expected.length=78 actual.length=74
	at org.junit.Assert.fail(Assert.java:93)
	at org.junit.internal.ComparisonCriteria.assertArraysAreSameLength(ComparisonCriteria.java:70)
	at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:34)
	at org.junit.Assert.internalArrayEquals(Assert.java:416)
	at org.junit.Assert.assertArrayEquals(Assert.java:202)
	at org.junit.Assert.assertArrayEquals(Assert.java:215)
	at io.documentnode.epub4j.epub.ResourcesLoaderTest.verifyResources(ResourcesLoaderTest.java:178)
	at io.documentnode.epub4j.epub.ResourcesLoaderTest.testLoadResources_ZipFile(ResourcesLoaderTest.java:108)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:45)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:15)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:42)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:263)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:68)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:47)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:28)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:30)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:300)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:834)

File: epub4j-core/src/test/java/io/documentnode/epub4j/epub/ResourcesLoaderTest.java (hunk 4 above)
Comment (ag2s20150909) [reply epub4j-core/src/test/java/io/documentnode/epub4j/epub/ResourcesLoaderTest.java:28]: It's the same error on my computer when using unchanged code.

**Pull request closed without merging**